vm/os-windows.cpp

```diff
@@ -304,6 +304,54 @@ void handle_ctrl_c() {
   SetConsoleCtrlHandler(factor::ctrl_handler, TRUE);
 }
 
+static volatile bool stop_on_ctrl_break = false;
+static HANDLE ctrl_break_thread = NULL;
+
+static DWORD WINAPI ctrl_break_thread_proc(LPVOID mainThread) {
+  bool ctrl_break_handled = false;
+  while (true) {
+    if (factor::stop_on_ctrl_break) {
+      if (GetAsyncKeyState(VK_CANCEL) >= 0) { /* Ctrl-Break is released. */
+        ctrl_break_handled = false;  /* Wait for the next press. */
+      } else if (!ctrl_break_handled) {
+        /* This proc runs in its own thread without stopping the main
+           thread. Since in practice nobody uses the multi-VM stuff yet,
+           we just interrupt the first VM we can get. */
+        FACTOR_ASSERT(thread_vms.size() > 0);
+        THREADHANDLE thd = thread_vms.begin()->first;
+        factor_vm* vm = thread_vms.begin()->second;
+        /* Check if the VM thread has the same Id as the thread Id of the
+           currently active window. Note that thread Id is not a handle. */
+        DWORD fgThreadId = GetWindowThreadProcessId(GetForegroundWindow(),
+                                                    NULL);
+        if (!atomic::load(&vm->safepoint.fep_p)
+            && (fgThreadId == vm->thread_id)
+        ) {
+          atomic::store(&vm->skip_debugger_p, true);
+          vm->safepoint.enqueue_fep(vm);
+          wake_up_thread(thd); /* Try to wake up the thread. */
+          ctrl_break_handled = true;
+        }
+      }
+    }
+    Sleep(100);
+  }
+  return 0;
+}
+
+void factor_vm::primitive_disable_ctrl_break() {
+  stop_on_ctrl_break = false;
+}
+
+void factor_vm::primitive_enable_ctrl_break() {
+  stop_on_ctrl_break = true;
+  if (ctrl_break_thread == NULL) {
+    DisableProcessWindowsGhosting();
+    ctrl_break_thread = CreateThread(NULL, 0, factor::ctrl_break_thread_proc,
+                                     NULL, 0, NULL);
+  }
+}
+
 void lock_console() {}
```

Comment:
I think the polling thread should finish when ctrl+break is disabled. E.g your `while (true) ` should be `while (atomic::load(&stop_on_ctrl_break)) {. 

Also I only think you need one flag, the `vm->skip_debugger_p` one. When that flag is on, the loop is running and when off it terminates. More flags just complicate things.